src/Microsoft.Health.Dicom.Core/Features/Query/QueryParser.cs

```diff
@@ -120,7 +146,7 @@ private static void PostProcessFilterConditions(QueryExpressionImp parsedQuery)
             if (parsedQuery.FuzzyMatch == true)
             {
                 IEnumerable<QueryFilterCondition> potentialFuzzyConds = parsedQuery.FilterConditions
-                                                                            .Where(c => QueryLimit.IsValidFuzzyMatchingQueryTag(c.DicomTag)).ToList();
+                                                                            .Where(c => QueryLimit.IsValidFuzzyMatchingQueryTag(c.DicomTag, c.CustomTagFilterDetails?.VR)).ToList();
```

Comment:
Unfortunately the foreach modifies values so it needs to be a list!

---
In reply to: [586590074](https://github.com/microsoft/dicom-server/pull/567#discussion_r586590074) [](ancestors = 586590074)

Reply:
Ah, I see in the loop. I think we can refactor this then a bit to perform the replacements by enumerating over indices. Luckily we're only replacing elements instead of changing the count:

```csharp
IEnumerable<(StringSingleValueMatchCondition, int)> potentialFuzzyConds = parsedQuery.FilterConditions
    .Where(c => QueryLimit.IsValidFuzzyMatchingQueryTag(c.DicomTag, c.CustomTagFilterDetails?.VR))
    .Cast<StringSingleValueMatchCondition>()
    .Select((x, i) => (x, i));

foreach (var (cond, i) in potentialFuzzyConds)
{
    parsedQuery.FilterConditions[i] = new PersonNameFuzzyMatchCondition(cond.DicomTag, cond.Value);
}
```

Reply:
Unfortunately, that won't work either since parsedQuery.FilterConditions is still being modified!

Reply:
Ah! My mistake -- I didn't realize the internal [version was incremented on set too](https://github.com/dotnet/runtime/blob/f7de865465532de372f65b55e00a8ec2be1e5274/src/libraries/System.Private.CoreLib/src/System/Collections/Generic/List.cs#L162). My second one should still work since we avoid the enumerator